File: src/semantic_project_model/schedule.py
"""Readiness, delegation, and critical path."""

from __future__ import annotations

from dataclasses import dataclass
from typing import cast

from semantic_project_model.graph import adjacency, find_cycle, longest_path
from semantic_project_model.model import Entity, ProjectGraph

DIRECT_DELEGATION_SCORE = 75
REVIEW_DELEGATION_SCORE = 60
BOUNDED_SPIKE_SCORE = 40


@dataclass(frozen=True, slots=True)
class WorkAssessment:
    entity: Entity
    ready: bool
    blockers: tuple[str, ...]
    agentability: int
    recommendation: str


def _integer(value: object, default: int) -> int:
    return value if isinstance(value, int) and not isinstance(value, bool) else default


def _score(entity: Entity) -> int:
    raw = entity.attributes.get("delegation")
    if not isinstance(raw, dict):
        return 0
    delegation = cast(dict[str, object], raw)
    positive = sum(
        _integer(delegation.get(key), 0)
        for key in (
            "specification_completeness",
            "context_locality",
            "testability",
            "reversibility",
            "integration_independence",
        )
    )
    blast = _integer(delegation.get("blast_radius"), 5)
    return max(0, min(100, positive * 4 - blast * 3))


def _recommendation(entity: Entity, score: int) -> str:
    raw = entity.attributes.get("delegation")
    review = True
    if isinstance(raw, dict) and isinstance(raw.get("human_review"), bool):
        review = raw["human_review"]

    if score >= DIRECT_DELEGATION_SCORE and not review:
        return "delegate directly"
    if score >= REVIEW_DELEGATION_SCORE:
        return "delegate with review"
    if score >= BOUNDED_SPIKE_SCORE:
        return "bounded spike"
    return "human-led design"


def assess_work(project: ProjectGraph) -> tuple[WorkAssessment, ...]:
    work = {entity.id: entity for entity in project.by_kind("work_item")}
    complete = {"complete", "accepted", "superseded"}
    results: list[WorkAssessment] = []

    for entity_id, entity in sorted(work.items()):
        blockers = {
            relation.target_id
            for relation in project.outgoing(entity_id, {"blocks"})
            if relation.target_id in work and work[relation.target_id].status not in complete
        }
        blockers.update(
            relation.target_id
            for relation in project.outgoing(entity_id, {"requires"})
            if relation.target_id in project.entities
            and project.entities[relation.target_id].kind == "decision"
            and project.entities[relation.target_id].status not in complete
        )
        score = _score(entity)
        results.append(
            WorkAssessment(
                entity=entity,
                ready=entity.status in {"ready", "planned", "in_progress"} and not blockers,
                blockers=tuple(sorted(blockers)),
                agentability=score,
                recommendation=_recommendation(entity, score),
            )
        )
    return tuple(results)


def critical_path(project: ProjectGraph) -> tuple[str, ...]:
    work = {entity.id: entity for entity in project.by_kind("work_item")}
    edges = [
        (relation.target_id, relation.source_id)
        for relation in project.relations
        if relation.kind == "blocks" and relation.source_id in work and relation.target_id in work
    ]
    graph = adjacency(work, edges)
    if find_cycle(graph) is not None:
        return ()
    weights = {
        entity_id: _integer(entity.attributes.get("effort"), 1)
        for entity_id, entity in work.items()
    }
    return longest_path(graph, weights)

File: src/project-model/schedule.ts
import { adjacency, findCycle, longestPath } from "./graph.ts";
import { byKind, outgoing, type Entity, type ProjectGraph } from "./types.ts";

const DIRECT_DELEGATION_SCORE = 75;
const REVIEW_DELEGATION_SCORE = 60;
const BOUNDED_SPIKE_SCORE = 40;

export interface WorkAssessment {
  readonly entity: Entity;
  readonly ready: boolean;
  readonly blockers: ReadonlyArray<string>;
  readonly agentability: number;
  readonly recommendation: string;
}

const integer = (value: unknown, fallback: number): number =>
  typeof value === "number" && Number.isInteger(value) ? value : fallback;

const record = (value: unknown): Readonly<Record<string, unknown>> | undefined =>
  value !== null && typeof value === "object" && !Array.isArray(value)
    ? (value as Readonly<Record<string, unknown>>)
    : undefined;

const score = (entity: Entity): number => {
  const delegation = record(entity.attributes.delegation);
  if (delegation === undefined) return 0;
  const positive = [
    "specification_completeness",
    "context_locality",
    "testability",
    "reversibility",
    "integration_independence",
  ].reduce((total, key) => total + integer(delegation[key], 0), 0);
  const blast = integer(delegation.blast_radius, 5);
  return Math.max(0, Math.min(100, positive * 4 - blast * 3));
};

const recommendation = (entity: Entity, value: number): string => {
  const delegation = record(entity.attributes.delegation);
  const review =
    delegation !== undefined && typeof delegation.human_review === "boolean"
      ? delegation.human_review
      : true;
  if (value >= DIRECT_DELEGATION_SCORE && !review) return "delegate directly";
  if (value >= REVIEW_DELEGATION_SCORE) return "delegate with review";
  if (value >= BOUNDED_SPIKE_SCORE) return "bounded spike";
  return "human-led design";
};

export const assessWork = (project: ProjectGraph): ReadonlyArray<WorkAssessment> => {
  const work = new Map(byKind(project, "work_item").map((entity) => [entity.id, entity]));
  const complete = new Set(["complete", "accepted", "superseded"]);
  return [...work]
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([entityId, entity]) => {
      const blockers = new Set(
        outgoing(project, entityId, new Set(["blocks"]))
          .filter(
            (relation) =>
              work.has(relation.targetId) &&
              !complete.has(work.get(relation.targetId)!.status ?? ""),
          )
          .map((relation) => relation.targetId),
      );
      for (const relation of outgoing(project, entityId, new Set(["requires"]))) {
        const target = project.entities.get(relation.targetId);
        if (target?.kind === "decision" && !complete.has(target.status ?? "")) {
          blockers.add(relation.targetId);
        }
      }
      const agentability = score(entity);
      return {
        entity,
        ready:
          new Set(["ready", "planned", "in_progress"]).has(entity.status ?? "") &&
          blockers.size === 0,
        blockers: [...blockers].sort(),
        agentability,
        recommendation: recommendation(entity, agentability),
      };
    });
};

export const criticalPath = (project: ProjectGraph): ReadonlyArray<string> => {
  const work = new Map(byKind(project, "work_item").map((entity) => [entity.id, entity]));
  const edges = project.relations
    .filter(
      (relation) =>
        relation.kind === "blocks" && work.has(relation.sourceId) && work.has(relation.targetId),
    )
    .map((relation) => [relation.targetId, relation.sourceId] as const);
  const graph = adjacency(work.keys(), edges);
  if (findCycle(graph) !== undefined) return [];
  const weights = new Map(
    [...work].map(([entityId, entity]) => [entityId, integer(entity.attributes.effort, 1)]),
  );
  return longestPath(graph, weights);
};
